File: src/patterns/creational/abstract-factory/abstract-factory.go
package main

import "fmt"

// Component interfaces
type IChair interface {
	SitOn()
}

type ISofa interface {
	LieOn()
}

type ITable interface {
	PlaceOn()
}

// Abstract Factories
type IFurnitureFactory interface {
	CreateChair() IChair
	CreateSofa() ISofa
	CreateTable() ITable
}

// Concrete Factories
type VictorianFactory struct {

}

func (f *VictorianFactory) CreateChair() IChair {
	return &VictorianChair{}
}

func (f *VictorianFactory) CreateSofa() ISofa {
	return &VictorianSofa{}
}

func (f *VictorianFactory) CreateTable() ITable {
	return &VictorianTable{}
}

type ModernFactory struct {

}

func (f *ModernFactory) CreateChair() IChair {
	return &ModernChair{}
}

func (f *ModernFactory) CreateSofa() ISofa {
	return &ModernSofa{}
}

func (f *ModernFactory) CreateTable() ITable {
	return &ModernTable{}
}

// Components
type ModernChair struct {

}

func (c *ModernChair) SitOn(){
	fmt.Println("Sitting on a modern chair")
}

type VictorianChair struct {

}

func (c *VictorianChair) SitOn(){
	fmt.Println("Sitting on a victorian chair")
}

type ModernSofa struct {

}

func (s *ModernSofa) LieOn(){
	fmt.Println("Lying on a modern sofa")
}

type VictorianSofa struct {

}

func (s *VictorianSofa)LieOn() {
	fmt.Println("Lying on a victorian sofa")
}

type ModernTable struct {

}

func (t *ModernTable) PlaceOn() {
	fmt.Println("Placed an item on a modern table")
}

type VictorianTable struct {}

func (t *VictorianTable) PlaceOn() {
	fmt.Println("Placed an item on a victorian table")
}

func main(){
	modernFactory := &ModernFactory{}
	modernTable := modernFactory.CreateTable()

	modernTable.PlaceOn()
}

File: src/patterns/creational/abstract-factory/abstract-factory.ts
// Component interfaces (abstract components)

// Furniture Interfaces
interface Chair {
  hasLegs(): boolean;
  sitOn(): void;
}

interface Sofa {
  hasLegs(): boolean;
  lieOn(): void;
}

interface Table {
  hasLegs(): boolean;
  placeItems(): void;
}

// Factory Interface
interface FurnitureFactory {
  createChair(): Chair;
  createSofa(): Sofa;
  createTable(): Table;
}

// Concrete Factories

class VictorianFactory implements FurnitureFactory {
  createChair(): Chair {
    return new VictorianChair();
  }

  createTable(): Table {
    return new VictorianTable();
  }

  createSofa(): Sofa {
    return new VictorianSofa();
  }
}

class ModernFactory implements FurnitureFactory {
  createChair(): Chair {
    return new ModernChair();
  }

  createTable(): Table {
    return new ModernTable();
  }

  createSofa(): Sofa {
    return new ModernSofa();
  }
}

// Components
class VictorianChair implements Chair {
  hasLegs(): boolean {
    return true;
  }

  sitOn() {
    console.log("Sitting on a victorian chair");
  }
}

class ModernChair implements Chair {
  hasLegs(): boolean {
    return true;
  }

  sitOn() {
    console.log("Sitting on a modern chair");
  }
}

class VictorianSofa implements Sofa {
  hasLegs(): boolean {
    return true;
  }

  lieOn() {
    console.log("Lying on a victorian sofa");
  }
}

class ModernSofa implements Sofa {
  hasLegs(): boolean {
    return true;
  }

  lieOn() {
    console.log("Lying on a modern sofa");
  }
}

class VictorianTable implements Table {
  hasLegs(): boolean {
    return true;
  }

  placeItems(): void {
    console.log("Put an item on a victorian table");
  }
}

class ModernTable implements Table {
  hasLegs(): boolean {
    return true;
  }

  placeItems(): void {
    console.log("Put an item on a modern table");
  }
}

const victorianFactory = new VictorianFactory();
const victorianChair = victorianFactory.createChair();

victorianChair.sitOn();
